File: R/complexspectrogram-module.R
# ComplexSpectrogram wrapper
# Phase-preserving spectrogram analysis with complex FFT

#' ComplexSpectrogram
#'
#' Create a ComplexSpectrogram object from a Sound: a phase-preserving
#' spectrogram computed with a complex FFT.
#'
#' @param sound Sound object.
#' @param window_length Window length in seconds. Default 0.005.
#' @param maximum_frequency Maximum frequency to analyze, in Hz. Default 5000.
#' @return An object of class \code{ComplexSpectrogram} (a list with methods,
#'   dispatched via the shared \code{PraatObject} pattern).
#'
#' @examples
#' sound <- Sound$create_tone(frequency = 150, duration = 0.3)
#' cs <- ComplexSpectrogram(sound)
#' cs$get_amplitude(0.15, 150)
#'
#' @name ComplexSpectrogram
NULL

# ============================================================================
# Shared Method Dispatch Table
# ============================================================================

.complexspectrogram_methods <- new.env(hash = TRUE, parent = emptyenv())

# Validation
.complexspectrogram_methods$is_valid <- function(.self) .self$.cpp$is_valid()

# Time properties
.complexspectrogram_methods$xmin <- function(.self) .self$.cpp$get_xmin()
.complexspectrogram_methods$xmax <- function(.self) .self$.cpp$get_xmax()
.complexspectrogram_methods$nx <- function(.self) .self$.cpp$get_nx()
.complexspectrogram_methods$dx <- function(.self) .self$.cpp$get_dx()
.complexspectrogram_methods$x1 <- function(.self) .self$.cpp$get_x1()

# Frequency properties
.complexspectrogram_methods$ymin <- function(.self) .self$.cpp$get_ymin()
.complexspectrogram_methods$ymax <- function(.self) .self$.cpp$get_ymax()
.complexspectrogram_methods$ny <- function(.self) .self$.cpp$get_ny()
.complexspectrogram_methods$dy <- function(.self) .self$.cpp$get_dy()
.complexspectrogram_methods$y1 <- function(.self) .self$.cpp$get_y1()

# Query
.complexspectrogram_methods$get_amplitude <- function(.self, time, frequency) {
  .self$.cpp$get_amplitude(time, frequency)
}
.complexspectrogram_methods$get_phase <- function(.self, time, frequency) {
  .self$.cpp$get_phase(time, frequency)
}

# Conversion
.complexspectrogram_methods$to_sound <- function(.self, stretch_factor = 1.0) {
  sound_ptr <- .self$.cpp$to_sound(stretch_factor)
  Sound(.xptr = sound_ptr)
}
.complexspectrogram_methods$to_spectrogram <- function(.self) {
  spec_ptr <- .self$.cpp$to_spectrogram()
  Spectrogram(.xptr = spec_ptr)
}
.complexspectrogram_methods$to_spectrum <- function(.self, time) {
  spectrum_ptr <- .self$.cpp$to_spectrum(time)
  Spectrum(.xptr = spectrum_ptr)
}

lockEnvironment(.complexspectrogram_methods, bindings = TRUE)

# ============================================================================
# S3 Dispatch
# ============================================================================

#' @method $ ComplexSpectrogram
#' @export
`$.ComplexSpectrogram` <- function(x, name) {
  val <- .subset2(x, name)
  if (!is.null(val)) return(val)
  method <- .complexspectrogram_methods[[name]]
  if (is.null(method)) return(NULL)
  function(...) method(x, ...)
}

# ============================================================================
# Constructor
# ============================================================================

#' @export
ComplexSpectrogram <- function(sound, window_length = 0.005,
  maximum_frequency = 5000.0) {
  if (!inherits(sound, "Sound")) {
    stop("First argument must be a Sound object")
  }

  sound_ptr <- if (!is.null(sound$.xptr)) {
    sound$.xptr
  } else if (!is.null(sound$.cpp)) {
    sound$.cpp$ptr
  } else {
    stop("Cannot extract XPtr from Sound object")
  }

  cs_mod <- get_module("complexspectrogram_module")
  xptr <- cs_mod$complexspectrogram_create_from_sound(
    sound_ptr, window_length, maximum_frequency
  )
  cpp_obj <- cs_mod$RComplexSpectrogram$new(xptr)

  structure(list(
    .cpp = cpp_obj
  ), class = c("ComplexSpectrogram", "PraatObject"))
}

# ============================================================================
# S3 Methods
# ============================================================================

#' @export
as.data.frame.ComplexSpectrogram <- function(x, ...) {
  x$.cpp$as_data_frame()
}

#' @export
print.ComplexSpectrogram <- function(x, ...) {
  if (!x$is_valid()) {
    cat("Invalid ComplexSpectrogram object\n")
    return(invisible(x))
  }
  cat("ComplexSpectrogram:\n")
  cat(sprintf("  Time domain: [%.3f, %.3f] s\n", x$xmin(), x$xmax()))
  cat(sprintf("  Frequency domain: [%.1f, %.1f] Hz\n", x$ymin(), x$ymax()))
  cat(sprintf("  Time frames: %d (dx = %.6f s)\n", x$nx(), x$dx()))
  cat(sprintf("  Frequency bins: %d (df = %.3f Hz)\n", x$ny(), x$dy()))
  invisible(x)
}

File: src/modules/complexspectrogram_module.cpp
// complexspectrogram_module.cpp
// Rcpp Module for Praat ComplexSpectrogram object
// Part of the pladdrr package - Phase 2.4 extension

#include <Rcpp.h>
#include "module_common.h"

// Praat headers
#include "../praat.github.io/dwtools/ComplexSpectrogram.h"
#include "../praat.github.io/fon/Sound.h"
#include "../praat.github.io/fon/Spectrogram.h"
#include "../praat.github.io/fon/Spectrum.h"
#include "../praat.github.io/melder/melder.h"

using namespace Rcpp;

// Forward declaration - NUMfpp initialization
extern void NUMmachar();

// ============================================================================
// Free Functions for ComplexSpectrogram Creation
// ============================================================================

// Create ComplexSpectrogram from Sound
XPtr<structComplexSpectrogram> complexspectrogram_create_from_sound(
    XPtr<structSound> sound,
    double window_length,
    double maximum_frequency
) {
    if (sound.get() == nullptr) {
        Rcpp::stop("Invalid Sound object");
    }

    // Ensure NUMfpp is initialized
    NUMmachar();

    try {
        autoComplexSpectrogram cs = Sound_to_ComplexSpectrogram(
            sound.get(), window_length, maximum_frequency
        );
        return create_xptr_from_auto<structComplexSpectrogram>(cs);
    } catch (MelderError) {
        Melder_clearError();
        Rcpp::stop("Failed to create ComplexSpectrogram from Sound");
    }
}

// ============================================================================
// RComplexSpectrogram Module Class
// ============================================================================

class RComplexSpectrogram {
public:
    Rcpp::XPtr<structComplexSpectrogram> ptr;

    // ========================================================================
    // Constructors
    // ========================================================================

    // Default constructor (empty/invalid object)
    RComplexSpectrogram() : ptr(R_NilValue) {}

    // Constructor from external pointer
    RComplexSpectrogram(Rcpp::XPtr<structComplexSpectrogram> p) : ptr(p) {}

    // ========================================================================
    // Validation
    // ========================================================================

    bool is_valid() {
        return ptr.get() != nullptr;
    }

    // ========================================================================
    // Properties - Time
    // ========================================================================

    double get_xmin() {
        if (!is_valid()) return NA_REAL;
        return ptr->xmin;
    }

    double get_xmax() {
        if (!is_valid()) return NA_REAL;
        return ptr->xmax;
    }

    int get_nx() {
        if (!is_valid()) return 0;
        return static_cast<int>(ptr->nx);
    }

    double get_dx() {
        if (!is_valid()) return NA_REAL;
        return ptr->dx;
    }

    double get_x1() {
        if (!is_valid()) return NA_REAL;
        return ptr->x1;
    }

    // ========================================================================
    // Properties - Frequency
    // ========================================================================

    double get_ymin() {
        if (!is_valid()) return NA_REAL;
        return ptr->ymin;
    }

    double get_ymax() {
        if (!is_valid()) return NA_REAL;
        return ptr->ymax;
    }

    int get_ny() {
        if (!is_valid()) return 0;
        return static_cast<int>(ptr->ny);
    }

    double get_dy() {
        if (!is_valid()) return NA_REAL;
        return ptr->dy;
    }

    double get_y1() {
        if (!is_valid()) return NA_REAL;
        return ptr->y1;
    }

    // ========================================================================
    // Query Methods
    // ========================================================================

    double get_amplitude(double time, double frequency) {
        VALIDATE_PTR(ptr, ComplexSpectrogram);
        
        try {
            integer itime = Matrix_xToColumn(ptr, time);
            integer ifreq = Matrix_yToRow(ptr, frequency);
            
            if (itime < 1 || itime > ptr->nx || ifreq < 1 || ifreq > ptr->ny) {
                return NA_REAL;
            }
            
            return ptr->z[ifreq][itime];
        } catch (MelderError) {
            Melder_clearError();
            return NA_REAL;
        }
    }

    double get_phase(double time, double frequency) {
        VALIDATE_PTR(ptr, ComplexSpectrogram);
        
        try {
            integer itime = Matrix_xToColumn(ptr, time);
            integer ifreq = Matrix_yToRow(ptr, frequency);
            
            if (itime < 1 || itime > ptr->nx || ifreq < 1 || ifreq > ptr->ny) {
                return NA_REAL;
            }
            
            return ptr->phase[ifreq][itime];
        } catch (MelderError) {
            Melder_clearError();
            return NA_REAL;
        }
    }

    // ========================================================================
    // Conversion Methods
    // ========================================================================

    XPtr<structSound> to_sound(double stretch_factor) {
        VALIDATE_PTR(ptr, ComplexSpectrogram);
        
        try {
            autoSound sound = ComplexSpectrogram_to_Sound(ptr.get(), stretch_factor);
            return create_xptr_from_auto<structSound>(sound);
        } catch (MelderError) {
            Melder_clearError();
            Rcpp::stop("Failed to convert ComplexSpectrogram to Sound");
        }
    }

    XPtr<structSpectrogram> to_spectrogram() {
        VALIDATE_PTR(ptr, ComplexSpectrogram);
        
        try {
            autoSpectrogram spec = ComplexSpectrogram_to_Spectrogram(ptr.get());
            return create_xptr_from_auto<structSpectrogram>(spec);
        } catch (MelderError) {
            Melder_clearError();
            Rcpp::stop("Failed to convert ComplexSpectrogram to Spectrogram");
        }
    }

    XPtr<structSpectrum> to_spectrum(double time) {
        VALIDATE_PTR(ptr, ComplexSpectrogram);
        
        try {
            autoSpectrum spectrum = ComplexSpectrogram_to_Spectrum(ptr.get(), time);
            return create_xptr_from_auto<structSpectrum>(spectrum);
        } catch (MelderError) {
            Melder_clearError();
            Rcpp::stop("Failed to convert ComplexSpectrogram to Spectrum");
        }
    }

    DataFrame as_data_frame() {
        VALIDATE_PTR(ptr, ComplexSpectrogram);
        
        std::vector<double> times;
        std::vector<double> frequencies;
        std::vector<double> amplitudes;
        std::vector<double> phases;

        for (integer itime = 1; itime <= ptr->nx; itime++) {
            double time = Matrix_columnToX(ptr, itime);
            for (integer ifreq = 1; ifreq <= ptr->ny; ifreq++) {
                double freq = Matrix_rowToY(ptr, ifreq);
                times.push_back(time);
                frequencies.push_back(freq);
                amplitudes.push_back(ptr->z[ifreq][itime]);
                phases.push_back(ptr->phase[ifreq][itime]);
            }
        }

        return DataFrame::create(
            Named("time") = times,
            Named("frequency") = frequencies,
            Named("amplitude") = amplitudes,
            Named("phase") = phases
        );
    }
};

// ============================================================================
// Rcpp Module Definition
// ============================================================================

RCPP_MODULE(complexspectrogram_module) {
    using namespace Rcpp;

    class_<RComplexSpectrogram>("RComplexSpectrogram")
        // Constructors
        .constructor()
        .constructor<XPtr<structComplexSpectrogram>>()

        // Validation
        .method("is_valid", &RComplexSpectrogram::is_valid)

        // Properties - Time
        .method("get_xmin", &RComplexSpectrogram::get_xmin)
        .method("get_xmax", &RComplexSpectrogram::get_xmax)
        .method("get_nx", &RComplexSpectrogram::get_nx)
        .method("get_dx", &RComplexSpectrogram::get_dx)
        .method("get_x1", &RComplexSpectrogram::get_x1)

        // Properties - Frequency
        .method("get_ymin", &RComplexSpectrogram::get_ymin)
        .method("get_ymax", &RComplexSpectrogram::get_ymax)
        .method("get_ny", &RComplexSpectrogram::get_ny)
        .method("get_dy", &RComplexSpectrogram::get_dy)
        .method("get_y1", &RComplexSpectrogram::get_y1)

        // Query methods
        .method("get_amplitude", &RComplexSpectrogram::get_amplitude)
        .method("get_phase", &RComplexSpectrogram::get_phase)

        // Conversion
        .method("to_sound", &RComplexSpectrogram::to_sound)
        .method("to_spectrogram", &RComplexSpectrogram::to_spectrogram)
        .method("to_spectrum", &RComplexSpectrogram::to_spectrum)
        .method("as_data_frame", &RComplexSpectrogram::as_data_frame)
    ;
    
    // Factory functions (returns XPtr)
    function("complexspectrogram_create_from_sound", &complexspectrogram_create_from_sound);
}
